docs/architecture/adr-003-governance-proposal-deposit-auto-throttler.md

```diff
@@ -0,0 +1,161 @@
+Assume $n_{t_1}$ changed at a certain time $t_1$ (and $D_t$ was last updated at that time so it is equal to $D_{t_1}$), and has not changed up to a certain time $t_2$ when the deposit is queried or requested for a new proposal. At time $t_2$ all we need to know is $\tau = ticks = \lfloor \frac{\Delta t}{T} \rfloor = \lfloor \frac{t_2 - t_1}{T} \rfloor$ and  then a way to compute easily $D_{t_2}$ as a function of $D_{t_1}$ and the number of ticks elapsed $\tau$. Due to the fact that $n_t$ has not changed since $t_1$ and is still $n_{t_1}$, all we need is to do is apply the same rate of decrease/increase multiple times, i.e.
+
+$$
+D_{t+1} = \max(D_{\min}, D_t \times (1+ sign(n_t - N) \times \alpha \times \sqrt[k]{| n_t - N |})^{\tau})
+$$
+
+What does this mean in practice?
+
+1. The deposit value need to only be actually updated in state when $n_t$ changes, i.e. when proposals actually enter or exit the voting period. Time-based updates can be avoided, and the current value calculated on-demand. To further simplify the computation, the value of $1+\alpha \times \sqrt[k]{| n_t - N |}$ could also be computed at this time (and maybe cached by nodes in memory instead of being stored in blockchain state) since it also won’t change until $n_t$ also changes again.
+2. If the current deposit value at a certain time $t$ is requested - say $t_1$ was the last time the min deposit was actually updated in state due to a change of the total active proposals $n_t$, it can be computed lazily as
+$D_t = \max( D_{\min}, D_{t_1} \times \gamma^\tau)$, where again $\tau = \lfloor \frac{t - t_1}{T} \rfloor$ - i.e. how many ticks have passed since $t_1$, and $\gamma = 1+ sign(n_t - N) \times \alpha \times \sqrt[k]{| n_t - N |}$ 
+
+### Implementation
+
+The following parameters are added to the `x/gov` module params:
+
+- `min_deposit_floor`: floor value for the minimum deposit required for a proposal to enter the voting period
+- `min_deposit_update_period`: duration that dictates after how long the dynamic minimum deposit should be recalculated for time-based updates.
+- `target_active_proposals`:  the number of active proposals the dynamic minimum deposit should target.
+- `min_deposit_increase_ratio`: the ratio of increase for the minimum deposit when the number of active proposals exceeds the target by 1.
+- `min_deposit_decrease_ratio`: the ratio of increase for the minimum deposit when the number of active proposals is 1 less than the target.
+- `min_deposit_sensitivity_target_distance`: A positive integer representing the sensitivity of the dynamic minimum deposit increase/decrease to the distance from the target number of active proposals. The higher the number, the lower the sensitivity. A value of 1 represents the highest sensitivity.
```

Comment:
I would advocate for having a single struct handling all these related parameters. It makes even more sense considering we would have to duplicate all of them for the `MinInitialDeposit`.

Reply:
A separate struct for params is a deprecated method (see the other *Params structs) so that's why I didn't do it.

Reply:
I don't think the reason for the deprecation of the other `*Params` structs was related to the way the global `Params` struct should be defined.

In the [ADR](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-046-module-params.md) that deprecates x/params, it is clearly stated that :
> Developers can choose to either create a unique data structure for every field in Params or they can create a single Params structure as outlined above in the case of x/auth.

Looking at the [PR](https://github.com/cosmos/cosmos-sdk/pull/12631/files#diff-6459d5ff0e84a4b446613a13ddb888cceeaa22e3cf34eb6398c472b448cfc835) that deprecated x/params for the gov module, it is more likely that these structs were removed because they were not part of a global struct that could be used as the global `Params`. In fact, they were part of the `GenesisState` struct, so it was necessary to merge them into a new global `Params` struct.

For example, the `SendEnabled` field of the bank module param is a custom struct.

Reply:
Ok, this sounds good. I made the change on both code and adr